File: app/Quotes/c/src/data/Quote.c
// Copyright 10-Sept-2018 ºDeme
// GNU General Public License - V3 <http://www.gnu.org/licenses/>

#include "data/Quote.h"

/* .+.
-struct: Quote
  date: char *
  open: double
  close: double
  max: double
  min: double
  vol: int
  error: bool
*/
/*.-.*/
#include "dmc/ct/Ajson.h"

struct quote_Quote {
  char *date;
  double open;
  double close;
  double max;
  double min;
  int vol;
  bool error;
};

Quote *quote_new(
  char *date,
  double open,
  double close,
  double max,
  double min,
  int vol,
  bool error
) {
  Quote *this = MALLOC(Quote);
  XNULL(date)
  this->date = date;
  this->open = open;
  this->close = close;
  this->max = max;
  this->min = min;
  this->vol = vol;
  this->error = error;
  return this;
}

char *quote_date(Quote *this) {
  XNULL(this)
  return this->date;
}

double quote_open(Quote *this) {
  XNULL(this)
  return this->open;
}

double quote_close(Quote *this) {
  XNULL(this)
  return this->close;
}

double quote_max(Quote *this) {
  XNULL(this)
  return this->max;
}

double quote_min(Quote *this) {
  XNULL(this)
  return this->min;
}

int quote_vol(Quote *this) {
  XNULL(this)
  return this->vol;
}

bool quote_error(Quote *this) {
  XNULL(this)
  return this->error;
}
/*.-.*/

#define TY Quote
#define FN quote
#include "dmc/tpl/topt.c"
#include "dmc/tpl/tarr.c"
#undef TY
#undef FN

#define TY Aquote
#define FN aquote
#include "dmc/tpl/topt.c"
#undef TY
#undef FN


Quote *quote_from_str(char *q) {
  Achar *parts = str_csplit(q, ':');
  return quote_new(
    achar_get(parts, 0),
    atof(achar_get(parts, 1)),
    atof(achar_get(parts, 2)),
    atof(achar_get(parts, 3)),
    atof(achar_get(parts, 4)),
    atoi(achar_get(parts, 5)),
    str_eq(achar_get(parts, 6), "true") ? true : false
  );
}

char *quote_to_str(Quote *this) {
  return str_printf("%s:%.4f:%.4f:%.4f:%.4f:%d:%s",
    this->date,
    this->open,
    this->close,
    this->max,
    this->min,
    this->vol,
    this->error ? "true" : "false"
  );
}

File: app/MultiMarket/js/src/data/Quote.js
// Copyright 04-Jun-2019 ºDeme
// GNU General Public License - V3 <http://www.gnu.org/licenses/>

/** Nick */
export default class Quote {

  /* .
  _rc_ Quote : from
  date : string
  open : number
  close : number
  max : number
  min : number
  vol : number
  error : boolean
  */

  /*--*/
  /**
   * @param {string} date
   * @param {number} open
   * @param {number} close
   * @param {number} max
   * @param {number} min
   * @param {number} vol
   * @param {boolean} error
   */
  constructor (
    date,
    open,
    close,
    max,
    min,
    vol,
    error
  ) {

    /**
     * @private
     * @type {string}
     */
    this._date = date;

    /**
     * @private
     * @type {number}
     */
    this._open = open;

    /**
     * @private
     * @type {number}
     */
    this._close = close;

    /**
     * @private
     * @type {number}
     */
    this._max = max;

    /**
     * @private
     * @type {number}
     */
    this._min = min;

    /**
     * @private
     * @type {number}
     */
    this._vol = vol;

    /**
     * @private
     * @type {boolean}
     */
    this._error = error;

  }

  /**  @return {string} */
  get date () {
    return this._date;
  }

  /**  @return {number} */
  get open () {
    return this._open;
  }

  /**  @return {number} */
  get close () {
    return this._close;
  }

  /**  @return {number} */
  get max () {
    return this._max;
  }

  /**  @return {number} */
  get min () {
    return this._min;
  }

  /**  @return {number} */
  get vol () {
    return this._vol;
  }

  /**  @return {boolean} */
  get error () {
    return this._error;
  }

  /** @return {string}  */
  toString () {
    return `${this.date}:${this._open}:${this._close}:` +
           `${this._max}:${this._min}:${this._vol}:${this._error}`;
  }

  /**
   * @param {string} s
   * @return {!Quote}
   */
  static fromString (s) {
    const fs = s.split(":");
    return new Quote(
      fs[0], Number(fs[1]), Number(fs[2]), Number(fs[3]),
      Number(fs[4]), Number(fs[5]), fs[6] === "true"
    );
  }

  /**
   * @param {!Array<?>} serial
   * @return {!Quote}
   */
  static fromJs (serial) {
    return new Quote(
      serial[0],
      serial[1],
      serial[2],
      serial[3],
      serial[4],
      serial[5],
      serial[6]
    );
  }
  /*--*/
}
